drivers/SmartThings/jbl/src/lunchbox/rest.lua

```diff
@@ -0,0 +1,403 @@
+
+    if headers:get_one("Content-Length") then
+      full_response = recv_additional_response(initial_recv, sock)
+    elseif headers:get_one("Transfer-Encoding") == "chunked" then
+      full_response = parse_chunked_response(initial_recv, sock)
+    else
+      full_response = initial_recv
+    end
+
+    return full_response
+  else
+    return nil, initial_err, partial
+  end
+end
+
+local function execute_request(client, request, retry_fn)
+  if client.socket == nil then
+    local success, err = connect(client)
+    if not success then return nil, err end
+  end
+
+  local should_retry = retry_fn
+
+  if type(should_retry) ~= "function" then
+    should_retry = function() return false end
+  end
+
+  -- send output
+  local _bytes_sent, send_err, _idx = nil, nil, 0
+  -- recv output
+  local response, recv_err, _partial = nil, nil, nil
+  -- return values
+  local ret, err = nil, nil
+
+  local backoff = backoff_builder(60, 1, 0.1)
+  local current_state = RestCallStates.SEND
+
+  repeat
+    local retry = should_retry()
+    if current_state == RestCallStates.SEND then
+      backoff = backoff_builder(60, 1, 0.1)
+      _bytes_sent, send_err, _idx = send_request(client, request)
+
+      if not send_err then
+        current_state = RestCallStates.RECEIVE
+      elseif retry then
+        if string.lower(send_err) == "closed" or string.lower(send_err):match("broken pipe") then
+          current_state = RestCallStates.RECONNECT
+        else
+          current_state = RestCallStates.RETRY
+        end
+      else
+        ret = nil
+        err = send_err
+        current_state = RestCallStates.COMPLETE
+      end
+    elseif current_state == RestCallStates.RECEIVE then
+      response, recv_err, _partial = handle_response(client.socket)
+
+      if not recv_err then
+        ret = response
+        err = nil
+        current_state = RestCallStates.COMPLETE
+      elseif retry then
+        if string.lower(recv_err) == "closed" or string.lower(recv_err):match("broken pipe") then
+          current_state = RestCallStates.RECONNECT
+        else
+          current_state = RestCallStates.RETRY
+        end
+      else
+        ret = nil
+        err = recv_err
+        current_state = RestCallStates.COMPLETE
+      end
+    elseif current_state == RestCallStates.RECONNECT then
+      local success, reconn_err = reconnect(client)
+      if success then
+        current_state = RestCallStates.RETRY
+      elseif not retry then
+        ret = nil
+        err = reconn_err
+        current_state = RestCallStates.COMPLETE
+      else
+        socket.sleep(backoff())
+      end
+    elseif current_state == RestCallStates.RETRY then
+      bytes_sent, send_err, _idx = nil, nil, 0
+      response, recv_err, partial = nil, nil, nil
+      current_state = RestCallStates.SEND
+      socket.sleep(backoff())
+    end
+  until current_state == RestCallStates.COMPLETE
+
+  return ret, err
+end
+
+local function make_socket(host, port, wrap_ssl)
```

Comment:
I would look at the updated version of this in Hue or Sonos, as there's a good bit of error handling that was missing that we added recently.

Reply:
update this file to latest version in sample